### ISSUE TITLE
Not working today

### ISSUE THREAD

**Issue by davilresende** [kelciour/awesometts-anki-addon#47]
Hello! It seems the program is not working today. I can't even see the speaker icon when I'm adding a new card.

**Comment by davilresende** [comment]
![image](https://user-images.githubusercontent.com/80525103/110941154-a07d0880-8316-11eb-86ec-321988ef3be9.png)

**Comment by kelciour** [comment]
The current version doesn't work well with the latest Anki version (2.1.41+) and for the time being is limited to 2.1.40 as the latest supported version.

After using the add-on from the Editor screen (Ctrl+T), Anki will be still responsive but will use a lot of CPU in the background and can only be closed with the Task Manager.

Maybe the add-on will be updated over the weekend or some time next week to support Anki 2.1.41+, though I'm not sure.

**Comment by lovac42** [comment]
It's probably caused by this line here: https://github.com/kelciour/awesometts-anki-addon/blob/master/awesometts/gui/generator.py#L760-L761

```
            while self.callback_message is None:
                app.instance().processEvents()
```
Probably some code was moved to background process in recent versions and it went into an infinite loop because the signal was called in a different thread.

It's just trying to strip the html and return the inner texts from the field. Rewriting it in pure python would be one way to fix it.

Format:
```
[ selected_html, selected_text, clipboard_html, clipboard_text ]
```

Here is the original 2.0 version:
```
        for origin in [
                lambda: from_note(web.selectedText()),
                lambda: from_note(web.page().mainFrame().evaluateJavaScript(
                    # for jQuery, this needs to be html() instead of text() as
                    # $('<div>hi<br>there</div>').text() yields "hithere"
                    # whereas if we have the original HTML, we can convert the
                    # line break tag into whitespace during input sanitization
                    '$("#f%d").html()' % editor.currentField
                )),
                lambda: try_clipboard('html'),
                lambda: try_clipboard('text'),
        ]:
```

**Comment by kelciour** [comment]
@lovac42 Thanks for looking into it! You're right. When this code was written, while trying to replace sync evaluateJavaScript with async runJavaScript using https://riverbankcomputing.com/pipermail/pyqt/2015-January/035324.html,  I assumed that the JavaScript callback will always return some string and it'd be safe to just check for `None` in the while loop to see if the callback was received or not.

I just tried without JS, it worked most of the time, but after making some changes to the field's content and immediately pressing Ctrl+T, the AwesomeTTS might pick up the outdated value.

```python
        def try_field():
            return editor.note.fields[editor.currentField]

        for origin in [
                lambda: ...
                lambda: from_note(try_field()),
                lambda: ...
```

In the end, the code was kept as is and just updated to keep up with Anki changes and a new variable was added to explicitly check for js callback.

@davilresende The issue should be fixed now.
